mipengine/controller/api/specifications_dtos.py

```diff
@@ -216,27 +236,33 @@ def _convert_transformer_specification_to_dto(spec: AlgorithmSpecification):
             if spec.parameters
             else None
         ),
-        compatible_algorithms=spec.compatible_algorithms,
+        preprocessing=_convert_transformers_specs_to_algorithm_preprocessing(
+            spec.name, transformers
+        ),
```

Comment:
This is not about who cares about what, it is about having to maintain 2 different models for the same concept, in the same system. I do not disagree on having 2 separate classes for the concept `AlgorithmSpecification`/`TransformerSpecification` and `AlgorithmSpecificationDTO`/`TransformerSpecificationDTO`. What I am saying is that having each one of them implementing the same concept in a different way(different model) is bad. The 2 different ways are:
1. In the `AlgorithmSpecification`/`TransformerSpecification` realm, the transformer-algorithm compatibility information is stored in `TransformerSpecification` class
2. In `AlgorithmSpecificationDTO`/`TransformerSpecificationDTO` realm, the transformer-algorithm compatibility information is stored in `AlgorithmSpecification` class

_"The problem with the model alteration that you mentioned showcases exactly that point. The algorithm developer should not care about transformers when he is creating a new algorithm but the "transformer developer" should care."_
When a new algorithm is added or when a new transformer is added one also needs to take care to add the compatibility information of the existing algorithms and transformers. If this is going to be stored in some algorithm specification or some transformer specification makes little difference, the important things is to be consistent about it, not storing it on a transformer specification but then convert it to be stored in an algorithm specification or vice versa. Otherwise, all these conversions from one to the other, create a serious mental burden for no reason.

_"On the other hand, when someone runs an algorithm in the engine, he mainly cares about the algorithm and the transformer is just some preprocessing to achieve proper algorithm execution."_
This is something to deal with outside the system. The system allows execution of algorithms with or without transformers, how a user makes the request is a UI issue, irrelevant to the internals of the system.

Reply:
This was the specification and the initial design of the longitudinal transformation.

https://team-1617704806227.atlassian.net/browse/MIP-749

The endpoint was suggested to be designed that way for ease of use, from the frontend perspective, and to make it more clear on how the algorithm request should be made. Otherwise we would have a /algorithms endpoint and a /transformers endpoint, but the algorithm request would be made only to the /algorithms endpoint including the transformers parameters which is much worse from my perspective. We would make the life harder for someone integrating, just to have our code a bit cleaner.

Having the simplification of the endpoint as a priority, we could not keep the internal models consistent. The specification of the transformer is separate and cannot be stored inside each algorithm specification, because that would create a lot of duplication at best.

Reply:
I do not see anything in the "Longitudinal data transformation" ticket referring to how the transformers and algorithms specifications are communicated to the caller of the system, but whatever...

_"Having the simplification of the endpoint as a priority, we could not keep the internal models consistent. The specification of the transformer is separate and cannot be stored inside each algorithm specification, because that would create a lot of duplication at best."_
Prioritizing the simplification of an endpoint over adding complexity to an already complex system makes no sense to me but whatever... 

_"The specification of the transformer is separate and cannot be stored inside each algorithm specification,"_
You would not need to store the specification of the transformer inside each algorithm specification. You would still have seperate algorithm and transformer specifications. The only debatable information here is the **_transformer-algorithm compatibility information_**. The `compatible_algorithms` field from the transformer specification should be removed, instead have a `compatible_transformers` in each algorithm specification, change all the relevant mechanisms accordingly and there you have the consistency between  `AlgorithmSpecification`/`TransformerSpecification` and `AlgorithmSpecificationDTO`/`TransformerSpecificationDTO`. This would be the correct way to do it, but again... whatever...